calyx-py/calyx/queues.py

```diff
@@ -151,12 +144,360 @@ def peek(self) -> Optional[int]:
+
+    - If a value is passed in, we pop the first (lowest-rank) instance of that value which is 'ripe'.
+    - If no value is passed in but a time is,
+        we pop the first (lowest-rank) value that passes the predicate.
+
+    When asked to peek:
+    We do the same thing as `pop`, except:
+    - We peek into the PIEO instead of popping it - i.e. we don't remove any elements.
+
+    We compactly represent these similar operations through `query`, which takes in an additional
+    optional `remove` parameter (defaulted to False) to determine whether to pop or peek.
+    """
+    
+    def __init__(self, max_len: int):
+        """Initialize structure."""
+        self.max_len = max_len
+        self.data = []
+        self.insertion_count = 0
+
+    def ripe(self, element, time):
+        """Check that an element is 'ripe' - i.e. its ready time has passed"""
+        return element["time"] <= time
+    
+    def binsert(self, val, time, rank, l, r):
+        """Inserts element into list such that rank ordering is preserved
+        Uses variant of binary search algorithm.
+        """
+        if l == r:
+            return self.data.insert(l, {"val": val, "time": time, "rank": rank})
+
+        mid = (l + r) // 2
+
+        if rank == self.data[mid]["rank"]:
+            return self.data.insert(mid, {"val": val, "time": time, "rank": rank})
+
+        if rank > self.data[mid]["rank"]:
+            return self.binsert(val, time, rank, mid+1, r)
+
+        if rank < self.data[mid]["rank"]:
+            return self.binsert(val, time, rank, l, mid)
+        
+    def push(self, val, rank=0, time=0, insertion_count=None) -> None:
+        """Pushes to a PIEO.
+        Inserts element such that rank ordering is preserved
+        """
+        
+        # Breaks ties and maintains FIFO order (can pass either custom insertion order or use PIEO internal one).
+        # Left-shifts the rank 32 bits, before adding either a passed in `insertion_count` parameter or the internal one.
+        rank = (rank << 32) + (insertion_count or self.insertion_count)
+
+        #If there is no room left in the queue, raise an Overflow error
+        if len(self.data) == self.max_len:
+            raise QueueError("Overflow")
+        
+        #If there are no elements in the queue, or the latest rank is higher than all others, append to the end
+        if len(self.data) == 0 or rank >= self.data[len(self.data)-1]["rank"]:
+            self.data.append({"val": val, "time": time, "rank": rank})
+
+        #If the latest rank is lower than all others, insert to the front
+        elif rank <= self.data[0]["rank"]:
+            self.data.insert(0, {"val": val, "time": time, "rank": rank})
+
+        #Otherwise, use the log-time insertion function
+        else:
+            self.binsert(val, time, rank, 0, len(self.data))
+        
+        self.insertion_count += 1
+     
+    def query(self, time=0, val=None, remove=False, return_rank=False) -> Optional[int]:
```

Comment:
I understand this version of `query` a lot better!

Reply:
I think it's just easier to understand querying a sorted list than it is to understand querying a heap ... but it's a trade-off to make since I use a heap for the simpler structure.